[core][runtime_env] Always first do working_dir, and expose a env var for pip and conda (#42697)
ray-project/ray: doc/source/ray-core/handling-dependencies.rst

You can also specify files via a remote cloud storage URI; see :ref:`remote-uris` for details.

> If a `working_dir` is specified, it's always the first to be prepared, and it presents in the creation of other runtime envs in environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}`. This allows `pip` and `conda` to reference local files in the working_dir, e.g. `requirements.txt` or `environment.yml`.

Reviewer: ```suggestion
If you specify a `working_dir`, Ray always prepares it first, and it's present in the creation of other runtime environments in the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable. This sequencing allows `pip` and `conda` to reference local files in the `working_dir` like `requirements.txt` or `environment.yml`.
```

  - Example: ``{"packages":["tensorflow", "requests"], "pip_check": False, "pip_version": "==22.0.2;python_version=='3.8.11'"}``

  When specifying a path to a ``requirements.txt`` file, the file must be present on your local machine and it must be a valid absolute path or relative filepath relative to your local current working directory, *not* relative to the `working_dir` specified in the `runtime_env`.

>   Furthermore, referencing local files *within* a `requirements.txt` file is not directly supported (e.g., ``-r ./my-laptop/more-requirements.txt``, ``./my-pkg.whl``). Instead, a environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` is provided in the creation process. You can use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, as long as they're in the working_dir.

Reviewer: ```suggestion
  Furthermore, referencing local files *within* a `requirements.txt` file isn't directly supported (e.g., ``-r ./my-laptop/more-requirements.txt``, ``./my-pkg.whl``). Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable in the creation process. For example, use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, while ensuring they're in the `working_dir`.
```

  - Example: ``"pytorch_p36"``

  When specifying a path to a ``environment.yml`` file, the file must be present on your local machine and it must be a valid absolute path or a relative filepath relative to your local current working directory, *not* relative to the `working_dir` specified in the `runtime_env`.

>   Furthermore, referencing local files *within* a `environment.yml` file is not directly supported. Instead, a environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` is provided in the creation process.

Reviewer: ```suggestion
  Furthermore, referencing local files *within* a `environment.yml` file isn't directly supported. Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}`environment variable in the creation process.
```

Reviewer 2: Nit: Let's also include your full description from above, it's pretty helpful.

> Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable in the creation process. For example, use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, while ensuring they're in the `working_dir`.

(Or whatever is required to make it work for `conda`.  It's just helpful to have a more fleshed out example.)
